In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [ ]:
import pandas as pd
import statsmodels.formula.api as smf
import numpy as np
from tabulate import tabulate
from scipy import stats

In [ ]:
data = pd.read_csv("/content/drive/My Drive/Year 3 Sems 3/Data Analysis Project/Case Study/cleaned data.csv")

In [ ]:
data.head()

Suburb             Address  Rooms Type       Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h   707435.52     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1480000.00      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1035000.00      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u  1186997.35     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1465000.00     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0   3/9/2016       2.5    3067.0  ...       1.0  1.0     126.0          84.3   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0         194.2   
2   4/2/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3   4/2/2016       2.5    3067.0  ...       2.0  1.0       0.0          80.0   
4   4/3/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0     1900.0  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1     1900.0  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3     1900.0  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [ ]:
#add in column for log price
data["Log_Price"] = np.log(data["Price"])

# Convert the Date column to a datetime data type
data['Date'] = pd.to_datetime(data['Date'], format='%d/%m/%Y')

#add in column for year sold
data["YearSold"] = data["Date"].dt.year

#add in column for the month sold
data["MonthSold"] = data["Date"].dt.month

#compute the age of house
age = list()
for counter,year in enumerate(data["YearBuilt"]):
  #we take 2018 is because the dataset is collected at year 2018
  #and for any age that is negative, we straight away convert it to become 0
  tempAge = 2018 - year
  if tempAge < 0:
    age.append(0)
  else:
    age.append(tempAge)

#inserting house age into new column of dataframe
if "Age" not in data.columns:
  data.insert(len(data.columns), "Age", age)

#modify the space and dash in regionname column to be replaced with underscore
data["Regionname"] = data["Regionname"].str.replace(" ","_")
data["Regionname"] = data["Regionname"].str.replace("-","_")

In [ ]:
#perform one hot encoding onto type column
#to break categorical variable into a few columns
#This is for ease of comparison later on
data = pd.get_dummies(data,columns=["Type","Regionname"])

In [ ]:
data.head()

Suburb             Address  Rooms       Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0   707435.52     SS  Jellis   
1  Abbotsford        85 Turner St    2.0  1480000.00      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0  1035000.00      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0  1186997.35     VB  Rounds   
4  Abbotsford        5 Charles St    3.0  1465000.00     SP  Biggin   

        Date  Distance  Postcode  Bedroom2  ...  Type_t  Type_u  \
0 2016-09-03       2.5    3067.0       2.0  ...       0       0   
1 2016-12-03       2.5    3067.0       2.0  ...       0       0   
2 2016-02-04       2.5    3067.0       2.0  ...       0       0   
3 2016-02-04       2.5    3067.0       3.0  ...       0       1   
4 2017-03-04       2.5    3067.0       3.0  ...       0       0   

   Regionname_Eastern_Metropolitan  Regionname_Eastern_Victoria  \
0                                0                            0   
1                                0                            0   
2                                0                            0   
3                                0                            0   
4                                0                            0   

   Regionname_Northern_Metropolitan Regionname_Northern_Victoria  \
0                                 1                            0   
1                                 1                            0   
2                                 1                            0   
3                                 1                            0   
4                                 1                            0   

   Regionname_South_Eastern_Metropolitan  Regionname_Southern_Metropolitan  \
0                                      0                                 0   
1                                      0                                 0   
2                                      0                                 0   
3                                      0                                 0   
4                                      0                                 0   

   Regionname_Western_Metropolitan  Regionname_Western_Victoria  
0                                0                            0  
1                                0                            0  
2                                0                            0  
3                                0                            0  
4                                0                            0  

[5 rows x 34 columns]

In [ ]:
#list of independent variable
xV = ["Rooms","Distance","Bedroom2","Bathroom","Landsize","BuildingArea","MonthSold","YearSold","Age","Type_h","Type_t","Type_u","Propertycount","Regionname_Eastern_Metropolitan","Regionname_Eastern_Victoria","Regionname_Northern_Metropolitan","Regionname_Northern_Victoria","Regionname_South_Eastern_Metropolitan","Regionname_Southern_Metropolitan","Regionname_Western_Metropolitan","Regionname_Western_Victoria"]

#list of dependent variable
yV = ["Log_Price"]

#list of independent and dependent variable
allV = xV + yV

#linear equation
formula = 'Log_Price ~ Rooms + Bedroom2 + Distance + Bathroom + Landsize + BuildingArea + MonthSold + YearSold + Age + Type_h + Type_t + Type_u + Propertycount + Regionname_Eastern_Metropolitan + Regionname_Eastern_Victoria + Regionname_Northern_Metropolitan + Regionname_Northern_Victoria + Regionname_South_Eastern_Metropolitan + Regionname_Southern_Metropolitan + Regionname_Western_Metropolitan + Regionname_Western_Victoria'

print(allV)

['Rooms', 'Distance', 'Bedroom2', 'Bathroom', 'Landsize', 'BuildingArea', 'MonthSold', 'YearSold', 'Age', 'Type_h', 'Type_t', 'Type_u', 'Propertycount', 'Regionname_Eastern_Metropolitan', 'Regionname_Eastern_Victoria', 'Regionname_Northern_Metropolitan', 'Regionname_Northern_Victoria', 'Regionname_South_Eastern_Metropolitan', 'Regionname_Southern_Metropolitan', 'Regionname_Western_Metropolitan', 'Regionname_Western_Victoria', 'Log_Price']


In [ ]:
#visualize the correlation matrix
data[allV].corr()

Rooms  Distance  Bedroom2  Bathroom  \
Rooms                                  1.000000  0.183987  0.959815  0.385971   
Distance                               0.183987  1.000000  0.186249  0.094051   
Bedroom2                               0.959815  0.186249  1.000000  0.389636   
Bathroom                               0.385971  0.094051  0.389636  1.000000   
Landsize                               0.021793  0.032781  0.021919  0.032074   
BuildingArea                           0.075357  0.055844  0.073110  0.094365   
MonthSold                              0.012414  0.012044  0.013975  0.010213   
YearSold                               0.083815  0.231275  0.097915  0.044609   
Age                                   -0.024923 -0.238787 -0.028986 -0.080499   
Type_h                                 0.306185  0.216399  0.294561  0.146451   
Type_t                                -0.017912 -0.025583 -0.016828  0.039218   
Type_u                                -0.335331 -0.227355 -0.322901 -0.196045   
Propertycount                         -0.077061 -0.018135 -0.076733 -0.052004   
Regionname_Eastern_Metropolitan        0.113425  0.192591  0.114025  0.082014   
Regionname_Eastern_Victoria            0.018325  0.272960  0.019433  0.027989   
Regionname_Northern_Metropolitan      -0.119546 -0.186817 -0.117164 -0.148248   
Regionname_Northern_Victoria           0.025546  0.252682  0.026124  0.016949   
Regionname_South_Eastern_Metropolitan  0.037791  0.439855  0.040110  0.011249   
Regionname_Southern_Metropolitan      -0.009022 -0.253161 -0.010974  0.099348   
Regionname_Western_Metropolitan        0.019226 -0.018105  0.016690 -0.035438   
Regionname_Western_Victoria            0.013122  0.167196  0.013649 -0.000721   
Log_Price                              0.366856 -0.095942  0.353572  0.540207   

                                       Landsize  BuildingArea  MonthSold  \
Rooms                                  0.021793      0.075357   0.012414   
Distance                               0.032781      0.055844   0.012044   
Bedroom2                               0.021919      0.073110   0.013975   
Bathroom                               0.032074      0.094365   0.010213   
Landsize                               1.000000      0.417273   0.001197   
BuildingArea                           0.417273      1.000000  -0.001426   
MonthSold                              0.001197     -0.001426   1.000000   
YearSold                               0.022567      0.026723  -0.490577   
Age                                   -0.020496     -0.018528   0.046250   
Type_h                                 0.017529      0.053724   0.016517   
Type_t                                -0.012263     -0.009635  -0.015679   
Type_u                                -0.010812     -0.053994  -0.007112   
Propertycount                         -0.004854     -0.006733   0.003921   
Regionname_Eastern_Metropolitan        0.009869      0.016759   0.011507   
Regionname_Eastern_Victoria            0.017755      0.024696  -0.000160   
Regionname_Northern_Metropolitan      -0.007811     -0.026685  -0.007310   
Regionname_Northern_Victoria           0.024508      0.012607  -0.003446   
Regionname_South_Eastern_Metropolitan  0.016057      0.030639   0.019805   
Regionname_Southern_Metropolitan      -0.007418     -0.002568  -0.000942   
Regionname_Western_Metropolitan       -0.008008     -0.004825  -0.011020   
Regionname_Western_Victoria            0.001892     -0.002477   0.003904   
Log_Price                              0.028033      0.082866   0.011852   

                                       YearSold       Age    Type_h  ...  \
Rooms                                  0.083815 -0.024923  0.306185  ...   
Distance                               0.231275 -0.238787  0.216399  ...   
Bedroom2                               0.097915 -0.028986  0.294561  ...   
Bathroom                               0.044609 -0.080499  0.146451  ...   
Landsize                               0.022567 -

By referring to the result of correlation matrix table above, it shows that the relationship between Rooms and Bedroom2 is quite high, therefore we will choose either one variable from it. And we will choose Rooms variable due to having higher relationship with Log_Price variable which is our dependent variable

**Lee: start from here.**

In [ ]:
#remove bedroom2 variable
if "Bedroom2" in xV:
  xV.remove("Bedroom2")
  allV = xV + yV
  # Remove "+ Bedroom2" from the formula
  formula = formula.replace(" + Bedroom2", "")

In [ ]:
# Standardize beta coefficient method

# standardizing dataframe
  # select all columns in dataframe which contains only numeric value
  # drop the null value
  # and compute the relative z-score
df_z = data.select_dtypes(include=[np.number]).dropna().apply(stats.zscore)

# fitting regression
result = smf.ols(formula, data=df_z).fit()

# checking results
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:              Log_Price   R-squared:                       0.463
Model:                            OLS   Adj. R-squared:                  0.463
Method:                 Least Squares   F-statistic:                     1670.
Date:                Wed, 13 Sep 2023   Prob (F-statistic):               0.00
Time:                        07:07:21   Log-Likelihood:                -38612.
No. Observations:               34852   AIC:                         7.726e+04
Df Residuals:                   34833   BIC:                         7.742e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
                                            coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------

From the result above, we notice that p-values for landsize, propertycount and monthSold variable is not significant, therefore, we decide to drop it and re-run the analysis again.**(Lee 4.4)**

In [ ]:
if "Landsize" in xV:
  xV.remove("Landsize")
  allV = xV + yV
  # Remove "+ Landsize" from the formula
  formula = formula.replace(" + Landsize", "")

if "Propertycount" in xV:
  xV.remove("Propertycount")
  allV = xV + yV
  # Remove "+ Propertycount" from the formula
  formula = formula.replace(" + Propertycount", "")

if "MonthSold" in xV:
  xV.remove("MonthSold")
  allV = xV + yV
  # Remove "+ MonthSold" from the formula
  formula = formula.replace(" + MonthSold", "")

In [ ]:
# Standardize beta coefficient method

# standardizing dataframe
  # select all columns in dataframe which contains only numeric value
  # drop the null value
  # and compute the relative z-score
df_z = data.select_dtypes(include=[np.number]).dropna().apply(stats.zscore)

# fitting regression
result = smf.ols(formula, data=df_z).fit()

# checking results
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:              Log_Price   R-squared:                       0.463
Model:                            OLS   Adj. R-squared:                  0.463
Method:                 Least Squares   F-statistic:                     2003.
Date:                Wed, 13 Sep 2023   Prob (F-statistic):               0.00
Time:                        07:07:21   Log-Likelihood:                -38614.
No. Observations:               34852   AIC:                         7.726e+04
Df Residuals:                   34836   BIC:                         7.740e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                                            coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------

From the result above, it provides lots of valuable information for us to make informed decisions. R square of 0.463 indicates that 46.30% of variability in dependent variable can be explained by the independent variable. This shows a moderate relationship between independent variable and dependent variable. Since all the p-values of independent variable is less than 0.05, we can safely assume that all variable is significant. Apart from that, we can also come out with a linear equation to predict the y variable value based on multiple x variables.

Equation: Log_Price=(−7.18×10
−16
 )+(0.144⋅Rooms)+(−0.227⋅Distance)+(0.431⋅Bathroom)+(0.026⋅BuildingArea)+(0.108⋅Age)+(0.116⋅Type_h)+(0.013⋅Type_t)+(−0.141⋅Type_u)+(0.043⋅Regionname_Eastern_Metropolitan)+(0.039⋅Regionname_Eastern_Victoria)+(−0.098⋅Regionname_Northern_Metropolitan)+(0.017⋅Regionname_Northern_Victoria)+(0.052⋅Regionname_South_Eastern_Metropolitan)+(0.120⋅Regionname_Southern_Metropolitan)+(−0.105⋅Regionname_Western_Metropolitan)+(−0.026⋅Regionname_Western_Victoria)

**(Lee : 4.4)**

In [ ]:
#gather required information
mergedLst = list()
for i in range(len(xV)):
  lst = list((xV[i],round(result.pvalues[i+1],3),round(result.params[i+1],3),abs(round(result.params[i+1],3))))
  mergedLst.append(list(lst))

In [ ]:
#rank independent variable according to absolute standardize beta coefficient
sortedLst = sorted(mergedLst,key=lambda x:x[3],reverse=True)

In [ ]:
#tabulate the finding
col_names = ["Content","p-values","Coefficient","Absolute Coefficient"]
print(tabulate(sortedLst, headers=col_names, tablefmt="fancy_grid", showindex="always"))

╒════╤═══════════════════════════════════════╤════════════╤═══════════════╤════════════════════════╕
│    │ Content                               │   p-values │   Coefficient │   Absolute Coefficient │
╞════╪═══════════════════════════════════════╪════════════╪═══════════════╪════════════════════════╡
│  0 │ Bathroom                              │          0 │         0.431 │                  0.431 │
├────┼───────────────────────────────────────┼────────────┼───────────────┼────────────────────────┤
│  1 │ Distance                              │          0 │        -0.231 │                  0.231 │
├────┼───────────────────────────────────────┼────────────┼───────────────┼────────────────────────┤
│  2 │ Rooms                                 │          0 │         0.143 │                  0.143 │
├────┼───────────────────────────────────────┼────────────┼───────────────┼────────────────────────┤
│  3 │ Type_u                                │          0 │        -0.141 │                

The table shows the ranking of variable according to the standardized beta coefficient of independent variable(absolute coefficient). From this table, we get to knows that Bathroom variable is having the most significant effect onto the changes of dependent variable which is Price variable. Since it is having positive relationship, it means that as number of bathroom increase, the log price variable increase as well. Apart from that, distance variable also having great influence onto the log price variables. It shows us that as one house is getting close to the city, the log price is higher. For comparison among house type, it shows that type_u is having a higher coefficient value compare to other type. By analyzing the relationship, it tells us that unit house type is more cheaper than other house type. On the hand, if we compare among the region, it can clearly be seen that coefficient value of southern metropolitan is ahead of other region and since the relationship is positive, it means that the housing price at this area is more expensive than other area.**(Lee: Chapter 4.5)**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [ ]:
#perform machine learning to verify our computed r-squared
x_train, x_test, y_train, y_test = train_test_split(data[xV],data[yV], test_size = 0.20, random_state= 42)

In [ ]:
# Create a linear regression model
model = LinearRegression()

# Train the model on the training data
model.fit(x_train, y_train)

LinearRegression()

In [ ]:
# Make predictions on the test data
y_pred = model.predict(x_test)

In [ ]:
from sklearn.metrics import r2_score

# Calculate R2
r2 = r2_score(y_test, y_pred)
print(f"R2 Score: {r2}")

R2 Score: 0.4494157855919174


Since the r-squared that we computed from the OLS method which is 0.463 is close to the r-squared computed by the machine learning model, which is 0.449. We can assume that this model is well fit and therefore, we can verified that the computed linear regression is accurate.